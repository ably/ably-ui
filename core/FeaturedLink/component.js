(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else if(typeof exports === 'object')
		exports["AblyUi"] = factory();
	else
		root["AblyUi"] = root["AblyUi"] || {}, root["AblyUi"]["Core"] = root["AblyUi"]["Core"] || {}, root["AblyUi"]["Core"]["FeaturedLink"] = factory();
})(this, function() {
return /******/ (() => { // webpackBootstrap
/******/ 	"use strict";
/******/ 	var __webpack_modules__ = ({

/***/ "./src/core/FeaturedLink/component.css":
/*!*********************************************!*\
  !*** ./src/core/FeaturedLink/component.css ***!
  \*********************************************/
/***/ ((__unused_webpack_module, __webpack_exports__, __webpack_require__) => {

__webpack_require__.r(__webpack_exports__);
// extracted by mini-css-extract-plugin


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/make namespace object */
/******/ 	(() => {
/******/ 		// define __esModule on exports
/******/ 		__webpack_require__.r = (exports) => {
/******/ 			if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 				Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 			}
/******/ 			Object.defineProperty(exports, '__esModule', { value: true });
/******/ 		};
/******/ 	})();
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!********************************************!*\
  !*** ./src/core/FeaturedLink/component.js ***!
  \********************************************/
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _component_css__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./component.css */ "./src/core/FeaturedLink/component.css");

})();

__webpack_exports__ = __webpack_exports__.default;
/******/ 	return __webpack_exports__;
/******/ })()
;
});
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIndlYnBhY2s6Ly9BYmx5VWkuQ29yZS5bbmFtZV0vd2VicGFjay91bml2ZXJzYWxNb2R1bGVEZWZpbml0aW9uIiwid2VicGFjazovL0FibHlVaS5Db3JlLltuYW1lXS8uL3NyYy9jb3JlL0ZlYXR1cmVkTGluay9jb21wb25lbnQuY3NzPzE3MTgiLCJ3ZWJwYWNrOi8vQWJseVVpLkNvcmUuW25hbWVdL3dlYnBhY2svYm9vdHN0cmFwIiwid2VicGFjazovL0FibHlVaS5Db3JlLltuYW1lXS93ZWJwYWNrL3J1bnRpbWUvbWFrZSBuYW1lc3BhY2Ugb2JqZWN0Il0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBO0FBQ0E7QUFDQTtBQUNBO0FBQ0E7QUFDQTtBQUNBO0FBQ0E7QUFDQSx1Q0FBdUMsdURBQXVEO0FBQzlGLENBQUM7QUFDRCxPOzs7Ozs7Ozs7OztBQ1ZBOzs7Ozs7O1VDQUE7VUFDQTs7VUFFQTtVQUNBO1VBQ0E7VUFDQTtVQUNBO1VBQ0E7VUFDQTtVQUNBO1VBQ0E7VUFDQTtVQUNBO1VBQ0E7VUFDQTs7VUFFQTtVQUNBOztVQUVBO1VBQ0E7VUFDQTs7Ozs7V0N0QkE7V0FDQTtXQUNBO1dBQ0Esc0RBQXNELGtCQUFrQjtXQUN4RTtXQUNBLCtDQUErQyxjQUFjO1dBQzdELEUiLCJmaWxlIjoiY29yZS9GZWF0dXJlZExpbmsvY29tcG9uZW50LmpzIiwic291cmNlc0NvbnRlbnQiOlsiKGZ1bmN0aW9uIHdlYnBhY2tVbml2ZXJzYWxNb2R1bGVEZWZpbml0aW9uKHJvb3QsIGZhY3RvcnkpIHtcblx0aWYodHlwZW9mIGV4cG9ydHMgPT09ICdvYmplY3QnICYmIHR5cGVvZiBtb2R1bGUgPT09ICdvYmplY3QnKVxuXHRcdG1vZHVsZS5leHBvcnRzID0gZmFjdG9yeSgpO1xuXHRlbHNlIGlmKHR5cGVvZiBkZWZpbmUgPT09ICdmdW5jdGlvbicgJiYgZGVmaW5lLmFtZClcblx0XHRkZWZpbmUoW10sIGZhY3RvcnkpO1xuXHRlbHNlIGlmKHR5cGVvZiBleHBvcnRzID09PSAnb2JqZWN0Jylcblx0XHRleHBvcnRzW1wiQWJseVVpXCJdID0gZmFjdG9yeSgpO1xuXHRlbHNlXG5cdFx0cm9vdFtcIkFibHlVaVwiXSA9IHJvb3RbXCJBYmx5VWlcIl0gfHwge30sIHJvb3RbXCJBYmx5VWlcIl1bXCJDb3JlXCJdID0gcm9vdFtcIkFibHlVaVwiXVtcIkNvcmVcIl0gfHwge30sIHJvb3RbXCJBYmx5VWlcIl1bXCJDb3JlXCJdW1wiRmVhdHVyZWRMaW5rXCJdID0gZmFjdG9yeSgpO1xufSkodGhpcywgZnVuY3Rpb24oKSB7XG5yZXR1cm4gIiwiLy8gZXh0cmFjdGVkIGJ5IG1pbmktY3NzLWV4dHJhY3QtcGx1Z2luXG5leHBvcnQge307IiwiLy8gVGhlIG1vZHVsZSBjYWNoZVxudmFyIF9fd2VicGFja19tb2R1bGVfY2FjaGVfXyA9IHt9O1xuXG4vLyBUaGUgcmVxdWlyZSBmdW5jdGlvblxuZnVuY3Rpb24gX193ZWJwYWNrX3JlcXVpcmVfXyhtb2R1bGVJZCkge1xuXHQvLyBDaGVjayBpZiBtb2R1bGUgaXMgaW4gY2FjaGVcblx0dmFyIGNhY2hlZE1vZHVsZSA9IF9fd2VicGFja19tb2R1bGVfY2FjaGVfX1ttb2R1bGVJZF07XG5cdGlmIChjYWNoZWRNb2R1bGUgIT09IHVuZGVmaW5lZCkge1xuXHRcdHJldHVybiBjYWNoZWRNb2R1bGUuZXhwb3J0cztcblx0fVxuXHQvLyBDcmVhdGUgYSBuZXcgbW9kdWxlIChhbmQgcHV0IGl0IGludG8gdGhlIGNhY2hlKVxuXHR2YXIgbW9kdWxlID0gX193ZWJwYWNrX21vZHVsZV9jYWNoZV9fW21vZHVsZUlkXSA9IHtcblx0XHQvLyBubyBtb2R1bGUuaWQgbmVlZGVkXG5cdFx0Ly8gbm8gbW9kdWxlLmxvYWRlZCBuZWVkZWRcblx0XHRleHBvcnRzOiB7fVxuXHR9O1xuXG5cdC8vIEV4ZWN1dGUgdGhlIG1vZHVsZSBmdW5jdGlvblxuXHRfX3dlYnBhY2tfbW9kdWxlc19fW21vZHVsZUlkXShtb2R1bGUsIG1vZHVsZS5leHBvcnRzLCBfX3dlYnBhY2tfcmVxdWlyZV9fKTtcblxuXHQvLyBSZXR1cm4gdGhlIGV4cG9ydHMgb2YgdGhlIG1vZHVsZVxuXHRyZXR1cm4gbW9kdWxlLmV4cG9ydHM7XG59XG5cbiIsIi8vIGRlZmluZSBfX2VzTW9kdWxlIG9uIGV4cG9ydHNcbl9fd2VicGFja19yZXF1aXJlX18uciA9IChleHBvcnRzKSA9PiB7XG5cdGlmKHR5cGVvZiBTeW1ib2wgIT09ICd1bmRlZmluZWQnICYmIFN5bWJvbC50b1N0cmluZ1RhZykge1xuXHRcdE9iamVjdC5kZWZpbmVQcm9wZXJ0eShleHBvcnRzLCBTeW1ib2wudG9TdHJpbmdUYWcsIHsgdmFsdWU6ICdNb2R1bGUnIH0pO1xuXHR9XG5cdE9iamVjdC5kZWZpbmVQcm9wZXJ0eShleHBvcnRzLCAnX19lc01vZHVsZScsIHsgdmFsdWU6IHRydWUgfSk7XG59OyJdLCJzb3VyY2VSb290IjoiIn0=