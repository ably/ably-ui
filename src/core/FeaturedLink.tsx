import React, { CSSProperties, ReactNode } from "react";

import Icon from "./Icon";
import { ColorClass, ColorThemeSet } from "./styles/colors/types";
import cn from "./utils/cn";

type FeaturedLinkProps = {
  url: string;
  children: ReactNode;
  textSize?: string;
  iconColor?: ColorClass | ColorThemeSet;
  flush?: boolean;
  reverse?: boolean;
  additionalCSS?: string;
  newWindow?: boolean;
  onClick?: () => void;
  disabled?: boolean;
  /**
   * Optional class name for the icon.
   */
  iconClassName?: string;
};

type TargetProps = { target?: string; rel?: string };

// When generating links with target=_blank, we only add `noreferrer` to
// links that don't start with `/`, so we can continue tracking referrers
// across our own domains
const buildTargetAndRel = (url: string, newWindow: boolean) => {
  const props: TargetProps = {};

  if (newWindow) {
    props.target = "_blank";

    if (url.startsWith("/") && !url.startsWith("//")) {
      props.rel = "noopener";
    } else {
      props.rel = "noopenner noreferrer";
    }
  }

  return props;
};

const FeaturedLink = ({
  url,
  textSize = "text-p2",
  iconColor,
  flush = false,
  reverse = false,
  additionalCSS = "",
  newWindow = false,
  onClick = undefined,
  children,
  disabled = false,
  iconClassName = "",
}: FeaturedLinkProps) => {
  const targetAndRel = buildTargetAndRel(url, newWindow);

  return (
    <a
      {...(onClick ? {} : { href: url })}
      className={cn(
        "font-sans font-bold block group/featured-link",
        { "text-gui-unavailable pointer-events-none": disabled },
        {
          "text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-none focus-visible:outline-gui-focus":
            !disabled,
        },
        { "py-2": !flush },
        `ui-${textSize}`,
        additionalCSS,
      )}
      style={
        {
          "--featured-link-icon-size": `var(${textSize.replace(
            "text",
            "--fs",
          )})`,
        } as CSSProperties
      }
      {...targetAndRel}
      onClick={onClick}
      onKeyDown={(e) => {
        if ((e.key === "Enter" || e.key === " ") && onClick) {
          e.preventDefault();
          onClick();
        }
      }}
    >
      {reverse ? (
        <>
          <Icon
            name="icon-gui-arrow-long-right-outline"
            size={`calc(var(--featured-link-icon-size) * 1.25)`}
            color={iconColor}
            additionalCSS={cn(
              "align-middle mr-2 relative -top-px -right-1 transition-[right] transform rotate-180",
              { "group-hover/featured-link:right-0": !disabled },
              iconClassName,
            )}
          />
          {children}
        </>
      ) : (
        <>
          {children}
          <Icon
            name="icon-gui-arrow-long-right-outline"
            size={`calc(var(--featured-link-icon-size) * 1.25)`}
            color={iconColor}
            additionalCSS={cn(
              "align-middle ml-2 relative -top-px -left-1 transition-[left]",
              {
                "group-hover/featured-link:left-0": !disabled,
              },
              iconClassName,
            )}
          />
        </>
      )}
    </a>
  );
};

export default FeaturedLink;
