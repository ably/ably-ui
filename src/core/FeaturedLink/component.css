@layer components {
  .ui-featured-link {
    max-width: calc(100% - 1.25rem);
    @apply font-medium block;
    @apply text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus;
  }

  .ui-featured-link:hover svg {
    @apply ml-8;
  }

  .ui-featured-link-icon {
    transition: margin-left 100ms ease-in-out;
    @apply align-middle ml-4 relative -top-1;
  }
}
