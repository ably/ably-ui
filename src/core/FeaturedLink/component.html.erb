<%= link_to(@url, class: "font-sans font-bold block text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus group #{@text_size} #{@flush ? '' : 'py-8'}  #{@additional_css}", style: "--featured-link-icon-size: var(#{@text_size.gsub("text", "--fs")})") do %>
  <% if @reverse %>
    <%= render(AblyUi::Core::Icon.new(name: "icon-gui-link-arrow",
                                      size: "calc(var(--featured-link-icon-size) * 1.25)",
                                      color: @icon_color,
                                      additional_css: "align-middle mr-8 relative -top-1 -right-4 transition-all group-hover:right-0 transform rotate-180")) %>
    <%= content -%>
  <% else %>
    <%= content -%><%= render(AblyUi::Core::Icon.new(name: "icon-gui-link-arrow",
                                                     size: "calc(var(--featured-link-icon-size) * 1.25)",
                                                     color: @icon_color,
                                                     additional_css: "align-middle ml-8 relative -top-1 -left-4 transition-all group-hover:left-0")) %>
  <% end %>

<% end %>
