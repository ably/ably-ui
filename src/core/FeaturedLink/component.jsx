import React from "react";
import T from "prop-types";

import Icon from "../Icon/component.jsx";

const FeaturedLink = ({ url, textSize = "text-menu3", iconColor = "text-cool-black", flush = false, children }) => (
  <a
    href={url}
    className={`ui-featured-link ${textSize} ${flush ? "" : "py-8"} `}
    style={{ "--featured-link-icon-size": `var(${textSize.replace("text", "--fs")})` }}
  >
    {children}
    <Icon name="icon-gui-disclosure-arrow" size={`calc(var(--featured-link-icon-size) * 1.43)`} color={iconColor} additionalCSS="ui-featured-link-icon" />
  </a>
);

FeaturedLink.propTypes = {
  url: T.string,
  children: T.node,
  textSize: T.string,
  iconColor: T.string,
  flush: T.bool,
};

export default FeaturedLink;
