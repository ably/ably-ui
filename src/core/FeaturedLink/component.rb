module AblyUi
  module Core
    class FeaturedLink < ViewComponent::Base
      def initialize(
        url:,
        text_size: 'text-menu3',
        icon_color: 'ui-icon-cool-black',
        flush: false
      )
        @url = url
        @text_size = text_size
        @icon_color = icon_color
        @flush = flush
      end
    end
  end
end
