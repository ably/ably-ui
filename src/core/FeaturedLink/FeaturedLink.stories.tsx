import FeaturedLink from "../FeaturedLink";

export default {
  title: "JS Components/Featured Link",
  component: FeaturedLink,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  args: {
    url: "#",
    children: "Featured link",
  },
};

export const Default = {
  args: {},
};

export const Reverse = {
  args: {
    reverse: true,
  },
};

export const Large = {
  args: {
    textSize: "text-p1",
  },
};

export const Small = {
  args: {
    textSize: "text-p3",
  },
};

export const Pink = {
  args: {
    iconColor: "text-pink-500",
    additionalCSS: "text-pink-800",
  },
};
