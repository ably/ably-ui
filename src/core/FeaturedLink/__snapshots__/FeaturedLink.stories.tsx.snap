// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Featured Link Default smoke-test 1`] = `
<a href="#"
   class="font-sans font-bold block text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus group/featured-link ui-text-p2 py-8 "
   style="--featured-link-icon-size: var(--fs-p2);"
>
  Featured link
  <svg class="align-middle ml-8 relative -top-1 -left-4 transition-[left] group-hover/featured-link:left-0"
       style="width: calc(var(--featured-link-icon-size) * 1.25); height: calc(var(--featured-link-icon-size) * 1.25);"
  >
    <use xlink:href="#sprite-icon-gui-link-arrow">
    </use>
  </svg>
</a>
`;

exports[`Components/Featured Link Disabled smoke-test 1`] = `
<a href="#"
   class="font-sans font-bold block text-gui-unavailable pointer-events-none group/featured-link ui-text-p2 py-8 "
   style="--featured-link-icon-size: var(--fs-p2);"
>
  Featured link
  <svg class="align-middle ml-8 relative -top-1 -left-4 transition-[left] "
       style="width: calc(var(--featured-link-icon-size) * 1.25); height: calc(var(--featured-link-icon-size) * 1.25);"
  >
    <use xlink:href="#sprite-icon-gui-link-arrow">
    </use>
  </svg>
</a>
`;

exports[`Components/Featured Link Large smoke-test 1`] = `
<a href="#"
   class="font-sans font-bold block text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus group/featured-link ui-text-p1 py-8 "
   style="--featured-link-icon-size: var(--fs-p1);"
>
  Featured link
  <svg class="align-middle ml-8 relative -top-1 -left-4 transition-[left] group-hover/featured-link:left-0"
       style="width: calc(var(--featured-link-icon-size) * 1.25); height: calc(var(--featured-link-icon-size) * 1.25);"
  >
    <use xlink:href="#sprite-icon-gui-link-arrow">
    </use>
  </svg>
</a>
`;

exports[`Components/Featured Link Pink smoke-test 1`] = `
<a href="#"
   class="font-sans font-bold block text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus group/featured-link ui-text-p2 py-8 text-pink-800"
   style="--featured-link-icon-size: var(--fs-p2);"
>
  Featured link
  <svg class="text-pink-500 align-middle ml-8 relative -top-1 -left-4 transition-[left] group-hover/featured-link:left-0"
       style="width: calc(var(--featured-link-icon-size) * 1.25); height: calc(var(--featured-link-icon-size) * 1.25);"
  >
    <use xlink:href="#sprite-icon-gui-link-arrow">
    </use>
  </svg>
</a>
`;

exports[`Components/Featured Link Reverse smoke-test 1`] = `
<a href="#"
   class="font-sans font-bold block text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus group/featured-link ui-text-p2 py-8 "
   style="--featured-link-icon-size: var(--fs-p2);"
>
  <svg class="align-middle mr-8 relative -top-1 -right-4 transition-[right] group-hover/featured-link:right-0 transform rotate-180"
       style="width: calc(var(--featured-link-icon-size) * 1.25); height: calc(var(--featured-link-icon-size) * 1.25);"
  >
    <use xlink:href="#sprite-icon-gui-link-arrow">
    </use>
  </svg>
  Featured link
</a>
`;

exports[`Components/Featured Link Small smoke-test 1`] = `
<a href="#"
   class="font-sans font-bold block text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus group/featured-link ui-text-p3 py-8 "
   style="--featured-link-icon-size: var(--fs-p3);"
>
  Featured link
  <svg class="align-middle ml-8 relative -top-1 -left-4 transition-[left] group-hover/featured-link:left-0"
       style="width: calc(var(--featured-link-icon-size) * 1.25); height: calc(var(--featured-link-icon-size) * 1.25);"
  >
    <use xlink:href="#sprite-icon-gui-link-arrow">
    </use>
  </svg>
</a>
`;
